File: server/scrapper/impl_hoyolab.py
import json
from typing import Literal
from urllib.parse import SplitResult
from uuid import uuid4

import requests
from bs4 import BeautifulSoup

from server.utils import download_image
from server.utils.constant import USER_AGENT

from ..domain.PageType import PageType
from .AbsScrapper import AbsScrapper


class ImplHoyolab(AbsScrapper):
    def __init__(self) -> None:
        super().__init__()
        self.pageType = PageType.hoyolab

    def gen_args(self, url: SplitResult):
        args = super().gen_args(url)
        args.key = url.path.split("article")[-1][1:]
        return args

    async def scrap(self, args):
        apiPath = f"https://bbs-api-os.hoyolab.com/community/post/wapi/getPostFull?post_id={args.key}"
        req = requests.get(
            apiPath,
            headers={
                "x-rpc-language": "ko-kr",
                "x-rpc-timezone": "Asia/Seoul",
                "x-rpc-show-translated": "true",
            },
        )
        data = req.json()
        post = data["data"]["post"]

        userData = post["user"]
        postData = post["post"]

        title = postData["subject"]

        contentData = json.loads(postData["content"])
        img_list = contentData["imgs"]
        content = contentData["describe"]

        fname_list = []
        for img in img_list:
            fname = download_image(img, f"{uuid4()}")
            if fname is not None:
                fname_list.append(fname)
        return self._create_scrap(userData["uid"], userData["nickname"], args.key, f"{title}\n{content}", fname_list)


class ImplHoyoLink(ImplHoyolab):
    def preprocess_url(self, url: str):
        res = requests.get(url, headers={"User-Agent": USER_AGENT})
        meta_list = BeautifulSoup(res.content, features="html.parser").find_all("meta")
        item = None
        for it in meta_list:
            if it.get("content", "").startswith("0;url="):
                item = it
                break

        url = item.get("content")[7:-1]
        return super().preprocess_url(url)

    def gen_args(self, url) -> any:
        args = super().gen_args(url)
        args.key = url.fragment.split("article")[1].split("/")[1]
        return args


if __name__ == "__main__":
    tweet_url = input("Enter the Twitter URL: ")

    downloader = ImplHoyolab()
    result = downloader.scrap(tweet_url)
    print(result)

File: internal/scrapper/implHoyolab.go
package scrapper

import (
    "Scrapper/internal/dto"
    "Scrapper/internal/scrapper/util"
    "errors"
    "fmt"
    "github.com/goccy/go-json"
    "github.com/google/uuid"
    "io"
    "net/http"
    "net/url"
    "os/exec"
    "strings"
)

type implHoyolab struct {
    AbsScrapper
}

var _ Scrapper = (*implHoyolab)(nil)

func (h *implHoyolab) GenArgs(parsedURL *url.URL) ScrapArgs {
    args := h.AbsScrapper.GenArgs(parsedURL)
    var key string
    if parsedURL.Path == "/" {
        key = strings.Split(parsedURL.Fragment, "article/")[1]
    } else {
        key = strings.Split(parsedURL.Fragment, "article/")[1]
    }
    args.Key = strings.Split(key, "/")[0]

    return args
}

func (h *implHoyolab) Scrap(args *ScrapArgs) (dto.ScrapCreate, error) {
    apiPath := fmt.Sprintf("https://bbs-api-os.hoyolab.com/community/post/wapi/getPostFull?post_id=%s", args.Key)
    req, err := http.NewRequest("GET", apiPath, nil)
    if err != nil {
        return dto.ScrapCreate{}, err
    }
    req.Header.Set("x-rpc-language", "ko-kr")
    req.Header.Set("x-rpc-timezone", "Asia/Seoul")
    req.Header.Set("x-rpc-show-translated", "true")

    client := &http.Client{CheckRedirect: func(req *http.Request, via []*http.Request) error { return http.ErrUseLastResponse }}
    resp, err := client.Do(req)
    if err != nil {
        return dto.ScrapCreate{}, err
    }
    defer resp.Body.Close()

    var result map[string]interface{}
    body, _ := io.ReadAll(resp.Body)
    err = json.Unmarshal(body, &result)
    if err != nil {
        return dto.ScrapCreate{}, err
    }

    post := result["data"].(map[string]interface{})["post"].(map[string]interface{})
    userData := post["user"].(map[string]interface{})
    postData := post["post"].(map[string]interface{})

    title := postData["subject"].(string)
    contentData := postData["content"].(string)
    var content map[string]interface{}
    err = json.Unmarshal([]byte(contentData), &content)
    if err != nil {
        return dto.ScrapCreate{}, err
    }

    imgList := content["imgs"].([]interface{})

    contentText := content["describe"].(string)

    var fnameList []string
    for _, img := range imgList {
        uid := uuid.New()
        fname, err := util.DownloadImage(fmt.Sprint(img), uid.String(), h.config)
        if err == nil {
            fnameList = append(fnameList, fmt.Sprintf(fname))
        }
    }

    return h.createScrap(userData["uid"].(string), userData["nickname"].(string), args.Key, title+"\n"+contentText, fnameList), nil
}

type implHoyolink struct {
    implHoyolab
}

var _ Scrapper = (*implHoyolink)(nil)

func (h implHoyolink) PreprocessURL(urlPath string) (*url.URL, error) {
    cmd := exec.Command("curl", "-s", urlPath)
    output, err := cmd.Output()
    if err != nil {
        return nil, err
    }
    parts := strings.Split(string(output), "\"")
    if len(parts) <= 1 {
        return nil, errors.New("failed to Preprocess HoyoLink 1")
    }
    next := parts[1]

    cmd = exec.Command("curl", "-s", next)
    output, err = cmd.Output()
    parts = strings.Split(string(output), "\"")
    if len(parts) <= 1 {
        return nil, errors.New("failed to Preprocess HoyoLink 2")
    }
    next = parts[1]

    p := strings.Split(next, ";")[0]
    parsed, err := url.Parse(p)
    que, err := url.ParseQuery(parsed.RawQuery)
    next = que.Get("url")

    return h.implHoyolab.PreprocessURL(next)
}
